File: 3980-BestTimeToBuyAndSellStockUsingStrategy/3980-BestTimeToBuyAndSellStockUsingStrategy.py
# Last updated: 12/25/2025, 7:09:40 PM
class Solution:
    def maxProfit(self, prices: List[int], strategy: List[int], k: int) -> int:
        n = len(prices)

        prefix = [0]
        for i in range(n):
            prefix.append(prefix[-1] + prices[i])
            
        half_k = k // 2

        def sell(index):
            return prefix[index + half_k] - prefix[index]

        @cache
        def rec(index, used):
            if index >= n: return 0

            ans = rec(index + 1, used) + strategy[index] * prices[index]

            if not used and index + k <= n:
                ans = max(ans, rec(index + k, True) + sell(index+half_k))

            return ans

        ans = rec(0, False)
        rec.cache_clear()
        return ans

File: 3980-BestTimeToBuyAndSellStockUsingStrategy/3980-BestTimeToBuyAndSellStockUsingStrategy.cpp
// Last updated: 12/25/2025, 7:09:35 PM
class Solution {
public:
    long long maxProfit(vector<int>& prices, vector<int>& strategy, int k) {
        int n = prices.size();
        vector<long long> prefix(n+1, 0);
        long long best = 0;
        for (int i=0; i<n; i++) {
            prefix[i + 1] = prefix[i] + prices[i];
            best += prices[i] * strategy[i];
        }

        int half = k / 2;
        long long left = 0;
        long long right = 0;
        for (int i=k; i<n; i++) {
            right += (strategy[i] * prices[i]);
        }

        for (int i=0; i<n-k; i++) {
            best = max(best, left + (prefix[i + k] - prefix[i + half]) + right);
            left += (strategy[i] * prices[i]);
            right -= (strategy[i + k] * prices[i + k]);
        }
        best = max(best, left + (prefix[n] - prefix[n - half]) + right);
        
        return best;
    }
};